### PR TITLE
Feature/at nguyenha/20191224 java

### PR DESCRIPTION
**Pull request by at-nguyenha**
Java && Kotlin all Exercises in Intro and The Core

### CODE REVIEW

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/DarkWilderness.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    int knapsackLight(int value1, int weight1, int value2, int weight2, int maxW) {
+        int value = 0;
+        if ((weight1 <= maxW && (value1 > value2 || weight2 > maxW)) || (weight1 + weight2 <= maxW)) {
+            value += value1;
+            maxW -= weight1;
+        }
+        if (weight2 <= maxW) {
+            value += value2;
+            maxW -= weight2;
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/DarkWilderness.java]: ERROR: [STYLE][DLS_DEAD_LOCAL_STORE] Dead store to maxW in asiantech.internship.summer.intro.DarkWilderness.knapsackLight(int, int, int, int, int)
http://findbugs.sourceforge.net/bugDescriptions.html#DLS_DEAD_LOCAL_STORE

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/EruptionofLight.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+     * in the string appears at most as many times as
+     * the previous letter in the alphabet within the
+     * string; ie: b occurs no more times than a;
+     * c occurs no more times than b; etc.
+     * <p>
+     * Given a string, check whether it is beautiful.
+     */
+    static boolean isBeautifulString(String inputString) {
+        int[] m = new int[26];
+        for (int i : inputString.getBytes()) {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/EruptionofLight.java]: ERROR: [I18N][DM_DEFAULT_ENCODING] Found reliance on default encoding in asiantech.internship.summer.intro.EruptionofLight.isBeautifulString(String): String.getBytes()
http://findbugs.sourceforge.net/bugDescriptions.html#DM_DEFAULT_ENCODING

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/EdgeoftheOcean.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        }
+        for (int i = 0; i < size - 2; i++) {
+            if (sequence[i] >= sequence[i + 2]) {
+                check2 += 1;
+            }
+        }
+        if (check > 1) {
+            return false;
+        } else return check2 <= 1;
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/EdgeoftheOcean.java]: WARNING: [IfElseStmtsMustUseBraces] Avoid using if...else statements without curly braces
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#ifelsestmtsmustusebraces

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/EdgeoftheOcean.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    int matrixElementsSum(int[][] matrix) {
+        int sum = 0;
+        for (int i = 0; i < matrix.length; i++) {
+            for (int j = 0; j < matrix[0].length; j++) {
+                if (matrix[i][j] == 0) {
+                    for (int k = i; k < matrix.length; k++) {
+                        matrix[k][j] = 0;
+                    }
+                } else
+                    sum += matrix[i][j];
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/EdgeoftheOcean.java]: WARNING: [IfElseStmtsMustUseBraces] Avoid using if...else statements without curly braces
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#ifelsestmtsmustusebraces

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/EruptionofLight.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
+package asiantech.internship.summer.intro;
+
+public class EruptionofLight {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/EruptionofLight.java]: ERROR: [ClassNamingConventions] The utility class name 'EruptionofLight' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
+package asiantech.internship.summer.intro;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class ExploringtheWaters {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java]: ERROR: [ClassNamingConventions] The utility class name 'ExploringtheWaters' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        if (A.length != B.length) {
+            return false;
+        }
+        int countSwap = 0;
+        int[] copyA = Arrays.copyOf(A, A.length);
+        int[] copyB = Arrays.copyOf(B, B.length);
+
+        Arrays.sort(copyA);
+        Arrays.sort(copyB);
+        if (!Arrays.equals(copyA, copyB)) return false;
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#ifstmtsmustusebraces

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        for (int i = 0; i < A.length; i++) {
+            if (A[i] != B[i]) countSwap++;
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#ifstmtsmustusebraces

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/IslandofKnowledge.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
+package asiantech.internship.summer.intro;
+
+public class IslandofKnowledge {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/IslandofKnowledge.java]: ERROR: [ClassNamingConventions] The utility class name 'IslandofKnowledge' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
+package asiantech.internship.summer.intro;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class LandofLogic {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java]: ERROR: [ClassNamingConventions] The utility class name 'LandofLogic' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+     * is equal to product. If there is no such integer, return -1 instead.
+     */
+    static int digitsProduct(int product) {
+        for (int i = 1; i < 99999; i++) {
+            String digits = String.valueOf(i);
+            int result = (digits.charAt(0) - '0');
+            for (int j = 1; j < digits.length(); j++) {
+                result *= (digits.charAt(j) - '0');
+            }
+            if (result == product)
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#ifstmtsmustusebraces

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/SmoothSailing.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
+package asiantech.internship.summer.intro;
+
+import java.util.ArrayList;
+
+public class SmoothSailing {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/SmoothSailing.java]: ERROR: [ClassNamingConventions] The utility class name 'SmoothSailing' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/SmoothSailing.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /**
+     * Write a function that reverses characters in
+     * (possibly nested) parentheses in the input string.
+     */
+    static String reverseInParentheses(String inputString) {
+        StringBuilder tmpCh = new StringBuilder(new String(""));
+        StringBuilder tmpChRe = new StringBuilder(new String(""));
+        String tmp = new String("");
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/SmoothSailing.java]: ERROR: [StringInstantiation] Avoid instantiating String objects; this is usually unnecessary.
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_performance.html#stringinstantiation

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            if (inputString.charAt(i) != inputString.charAt(s - i - 1)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    public static void main(String[] args) {
+        System.out.println("Sum : " + add(1, 2));
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java]: ERROR: [SystemPrintln] System.out.println is used
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_bestpractices.html#systemprintln

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+        System.out.println("Century is : " + centuryFromYear());
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java]: ERROR: [SystemPrintln] System.out.println is used
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_bestpractices.html#systemprintln

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+        if (!checkPalindrome("abcddcba")) {
+            System.out.println("It is a palindrome");
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java]: ERROR: [SystemPrintln] System.out.println is used
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_bestpractices.html#systemprintln

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        } else {
+            System.out.println("It is not a palindrome");
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/TheJourneyBegin.java]: ERROR: [SystemPrintln] System.out.println is used
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_bestpractices.html#systemprintln

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ThroughtheFog.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
+package asiantech.internship.summer.intro;
+
+public class ThroughtheFog {
```
Comment (at-dev-support-01) [inline app/src/at-nguyenha/java/asiantech/internship/summer/intro/ThroughtheFog.java]: ERROR: [ClassNamingConventions] The utility class name 'ThroughtheFog' doesn't match '[A-Z][a-zA-Z0-9]+(Utils?|Helper)'
https://pmd.github.io/pmd-6.8.0/pmd_rules_java_codestyle.html#classnamingconventions

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java (hunk 7 above)
Comment (at-nguyenha) [reply app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java]: Done

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java (hunk 8 above)
Comment (at-nguyenha) [reply app/src/at-nguyenha/java/asiantech/internship/summer/intro/ExploringtheWaters.java]: Done

File: app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java (hunk 11 above)
Comment (at-nguyenha) [reply app/src/at-nguyenha/java/asiantech/internship/summer/intro/LandofLogic.java]: Done

**Pull request merged**